(In-Progress) Percent encode User ID

phildenhoff opened the pull request:
This closes: #211 

This PR changes the `path` parameter of `_build_path` to List[str], improves a number of tests, and fixes at least one.

Previously, accepting `path` as a "str" caused a bug where valid matrix names (like "@a-z0-9._=-/:example.com") were breaking as the "/" was interpreted as a path separation.
    
The benefit of moving to a List[str] is that each substring of the path can be separetly percent-encoded and then combined into one long path string. The above username becomes "%40a-z0-9._%3D-%2F%3Aexample.com" and there is no confusion whether that is a name or a path.

phildenhoff commented on tests/async_client_test.py:
```diff
@@ -749,6 +749,8 @@ async def test_sync(self, async_client, aioresponse):
 
         aioresponse.get(
+            # TODO: what the hell is the difference here?
+            #f'https://example.org/_matrix/client/r0/sync?access_token=abc123&since=s526_47314_0_7_1_1_1_11444_1&filter={quote("{}")}',
             re.compile(url + r"&filter=" + quote("{}") + "&since=.*"),
```
This is the test that isn't passing; the string above will work just fine, however.

poljar commented on nio/api.py:
```diff
@@ -234,11 +234,42 @@ def encrypted_mxc_to_plumb(mxc, key, hash, iv, homeserver=None):
+        )
+
+        # remove trailing /
+        if path[-1] == "/":
```
I don't think that check is necessary, `rstrip` will be a noop if the last char isn't a `/`, no?

poljar commented on tests/async_client_test.py:
```diff
@@ -749,6 +749,8 @@ async def test_sync(self, async_client, aioresponse):
         # Test with filter dict
 
         aioresponse.get(
+            # TODO: what the hell is the difference here?
```
I think it's fine to just use the more specific URL.

poljar replied:
I'm hitting that issue as well on master while migrating our CI over, seems like something is double quoting, though it happens only when using Tox so :shrug:

phildenhoff replied:
Okay, if you're happy using the specified string that works for me. My most recent commits changes that + removes the trailing slash check, and I fixed up the types casting.

poljar reviewed (commented):
What's the state of this, looking at it it seems to be fine?

phildenhoff commented:
Alright I rebased onto master and fixed up a few things. Your regex for that one test worked fine for me so I opted to use it instead. I notice there's no CI for push requests any more though

poljar reviewed (approved):
I fixed the github action to trigger on PRs as well, I can't seem to find how to retrigger one here but I did run the tests manually.

Thanks for the PR.
